rfcs/0006-cursor-paging.md

```diff
@@ -0,0 +1,164 @@
+# RFC0006 - Cursor-Based Paging
+
+* RFC Proposal ID: RFC0006 
+* Status: First Draft
+* Start Date: 2020-04-01
+* RFC PR: 
+* JIRA Ticket(s):
+    
+This RFC changes how the BFD handles requests for explanation of benefits (EOB). 
+Internally, it uses a more efficient database query to fetch EOBs. 
+For partners, it provides a way for BFD clients to request resources with lower timeouts and higher throughput.  
+
+## Motivation
+The latency of the BFD's EOB search is directly related to the number of claims a beneficiary has. 
+Since the number of claims per beneficiary varies widely, a BFD client must set 10 to 30-second timeout on their EOB fetches to accommodate a large number of claims in the result set. 
+
+Paging is the term given to the technique to split a single large result set into many smaller result sets. 
+Since smaller result sets take less processing time, paging should imply lower request latency. 
+The FHIR specification and the BFD API support paging. 
+Unfortunately, the BFD's current paging algorithm is an inefficient implementation of offset paging, where the BFD data server queries for the entire result set and then discards the results outside of the requested page. 
+
+The latency variability (also known as jitter) of a BFD EOB request leads to several problems: 
+- As mentioned, it requires the BFD client to set long timeout values
+- SLO and SLA latency metrics are difficult to specify because normal operations may take a long time to complete
+- Request retries, usually a good practice in distributed systems, can lead to a very wasteful failure-mode where computationally expensive partial results are repeatedly discarded. 
+- Jitter adversely affects the throughput of a system. 
+
+## Proposed Solution
+The proposal is to replace the current offset paging algorithm with a cursor-based algorithm. 
+Internally, the BFD can implement cursor-based paging with an efficient seek query. 
+From a database perspective, the work done is nearly constant for each request. 
```

Comment:
Can you please also describe this in terms of the SQL that would be executed, as compared to the current SQL queries? It's not clear to me if you're actually using DB cursors (i.e. <https://www.postgresql.org/docs/9.6/plpgsql-cursors.html>) or if you're just using seek-based paging. If you're using both, it's not clear why you feel the combination is important.

Reply:
This was captured in the unresolved questions section of the RFC as we accept this RFC under a _Deferred_ status.